frame/lottery/src/lib.rs

```diff
@@ -0,0 +1,298 @@
+}
+
+type Index = u32;
+
+// Any runtime call can be encoded into two bytes which represent the pallet and call index.
+// We use this to uniquely match someone's incoming call with the calls configured for the lottery.
+type CallIndex = (u8, u8);
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Lottery {
+		LotteryIndex: Index;
+		/// The configuration for the current lottery.
+		Lottery: Option<LotteryConfig<T::BlockNumber, BalanceOf<T>>>;
+		/// Users who have purchased a ticket. (Lottery Index, Tickets Purchased)
+		Participants: map hasher(twox_64_concat) T::AccountId => (Index, Vec<CallIndex>);
+		/// Total number of tickets sold.
+		TicketsCount: Index;
+		/// Each ticket's owner.
+		///
+		/// May have residual storage from previous lotteries. Use `TicketsCount` to see which ones
+		/// are actually valid ticket mappings.
+		Tickets: map hasher(twox_64_concat) Index => Option<T::AccountId>;
+	}
+}
+
+decl_event!(
+	pub enum Event<T> where
+		<T as frame_system::Trait>::AccountId,
+		Balance = BalanceOf<T>,
+		Call = <T as Trait>::Call,
+	{
+		/// A lottery has been started!
+		LotteryStarted,
+		/// A winner has been chosen!
+		Winner(AccountId, Balance),
+		/// A ticket has been bought!
+		TicketBought(AccountId, Call),
+	}
+);
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// An overflow has occurred.
+		Overflow,
+		/// A lottery has not been configured.
+		NotConfigured,
+		/// A lottery has not started.
+		NotStarted,
+		/// A lottery is already in progress.
+		InProgress,
+		/// A lottery has already ended.
+		AlreadyEnded,
+		/// The call is not valid for an open lottery.
+		InvalidCall,
+		/// You are already participating in the lottery with this call.
+		AlreadyParticipating,
+		/// Too many calls for a single lottery.
+		TooManyCalls,
+		/// Failed to encode calls
+		EncodingFailed,
+	}
+}
+
+#[derive(Encode, Decode, Default, RuntimeDebug)]
+pub struct LotteryConfig<BlockNumber, Balance> {
+	/// Index of this lottery
+	index: Index,
+	/// Price per entry.
+	price: Balance,
+	/// Starting block of the lottery.
+	start: BlockNumber,
+	/// Ending block of the lottery.
+	end: BlockNumber,
+	/// Payout block of the lottery.
+	payout: BlockNumber,
+	/// Calls available this lottery.
+	calls: Vec<CallIndex>,
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin, system = frame_system {
+		/// TODO: Expose Constants
+
+		fn deposit_event() = default;
+
+		#[weight = T::WeightInfo::setup_lottery(calls.len() as u32)]
+		fn setup_lottery(
+			origin,
+			price: BalanceOf<T>,
+			start: T::BlockNumber,
+			end: T::BlockNumber,
+			payout: T::BlockNumber,
+			calls: Vec<<T as Trait>::Call>,
+		) {
+			T::ManagerOrigin::ensure_origin(origin)?;
+			ensure!(calls.len() <= T::MaxCalls::get(), Error::<T>::TooManyCalls);
+			let indices = Self::calls_to_indices(&calls)?;
+			Lottery::<T>::try_mutate(|lottery| -> DispatchResult {
+				ensure!(lottery.is_none(), Error::<T>::InProgress);
```

Comment:
Why limit lotteries to one at a time?

Reply:
No reason specifically, would just be another feature and extended API.

Would need to think about how to check for multiple lotteries ending, holding multiple configs, holding multiple lottery pots, etc...